sofar/sofar.py

```diff
@@ -0,0 +1,2125 @@
+    """
+
+    # these have to be set here, because they are used in __setattr__ and
+    # Python checks if they exist upon class creation
+
+    # don't allow adding attributes and deleting/writing read only attributes
+    _protected = False
+    # list of read only attributes (filled upon init)
+    _read_only_attr = []
+
+    def __init__(self, convention, mandatory=False, version="latest",
+                 verify=True):
+        """See class docstring"""
+
+        # get convention
+        self._convention = self._load_convention(convention, version)
+
+        # update read only attributes
+        self._read_only_attr = [
+            key for key in self._convention.keys()
+            if self._read_only(self._convention[key]["flags"])]
+
+        # add attributes with default values
+        self._convention2sofa(mandatory)
+
+        # add and update the API
+        if verify:
+            self.verify(version)
+
+        self._protected = True
+
+    def __setattr__(self, name: str, value):
+        # don't allow new attributes to be added outside the class
+        if self._protected and not hasattr(self, name):
+            raise TypeError(f"{name} is an invalid attribute")
+
+        # don't allow setting read only attributes
+        if name in self._read_only_attr and self._protected:
+            raise TypeError(f"{name} is a read only attribute")
+
+        # convert to numpy array or scalar
+        if not isinstance(value, (str, dict, np.ndarray)):
+            value = np.atleast_2d(value)
+            if value.size == 1:
+                value = value.flatten()[0]
+
+        super.__setattr__(self, name, value)
+
+    def __delattr__(self, name: str):
+        # can't delete non existing attributes
+        if not hasattr(self, name):
+            raise TypeError(f"{name} is not an attribute")
+        # delete anything if not frozen, delete non mandatory
+        if not self._protected or \
+                not self._mandatory(self._convention[name]["flags"]):
+            super().__delattr__(name)
+
+            # check if custom field as to be deleted
+            if hasattr(self, "_custom"):
+                if name in self._custom:
+                    self._custom.pop(name)
+        else:
+            raise TypeError(
+                f"{name} is a mandatory attribute that can not be deleted")
+
+    def __repr__(self):
+        return (f"sofar.SOFA object: {self.GLOBAL_SOFAConventions} "
+                f"{self.GLOBAL_SOFAConventionsVersion}")
+
+    @property
+    def dimensions(self):
+        """
+        Print the dimensions of the SOFA object
+
+        The SOFA file standard defines the following dimensions:
+
+        M
+            number of measurements
+        N
+            number of samles, frequencies, SOS coefficients, SH coefficients
+            (depending on self.GLOBAL_DataType)
+        R
+            Number of receivers
+        E
+            Number of emitters
```

Comment:
Need to reflect the SH coefficient specifications

Reply:
yes, good point - added that

Reply:
yes, done :)